src/routes/register/index.js

```diff
@@ -1,88 +1,184 @@
-		if (await this.validate()) {
-			this.props.onSubmit(Array.from(this.state.fields).map((el) => [el.props.name, el.value]).reduce((values, [key, value]) => ({ ...values, [key]: value }), {}));
+	state = {
+		name: null,
+		email: null,
+		department: null,
+	}
+
+	validations = {
+		name: [Validations.nonEmpty],
+		email: [Validations.nonEmpty, Validations.email],
+		department: [Validations.nonEmpty],
+	}
+
+	validate = (fieldName, value) => this.validations[fieldName].reduce((error, validation) => (error || validation(value)), undefined)
+
+	validateAll = () => {
+		for (const fieldName of Object.keys(this.validations)) {
+			const { value } = this.state[fieldName];
+			const error = this.validate(fieldName, value);
+			this.setState({ [fieldName]: { ...this.state[fieldName], value, error, showError: false } });
 		}
 	}
 
-	async validate() {
-		const valid = await asyncEvery(Array.from(this.state.fields), async (el) => await el.validate());
-		this.setState({
-			valid,
-		});
-		return valid;
+	handleFieldChange = (fieldName) => ({ target: { value } }) => {
+		const error = this.validate(fieldName, value);
+		this.setState({ [fieldName]: { ...this.state[fieldName], value, error, showError: false } });
 	}
 
-	addToValidate(element) {
-		this.state.fields.add(element);
+	handleNameChange = this.handleFieldChange('name')
+
+	handleEmailChange = this.handleFieldChange('email')
+
+	handleDepartmentChange = this.handleFieldChange('department')
+
+	handleSubmit = (event) => {
+		event.preventDefault();
+
+		if (this.props.onSubmit) {
+			const values = Object.entries(this.state)
+				.map(([name, { value }]) => ({ [name]: value }))
+				.reduce((values, entry) => ({ ...values, ...entry }), {});
+			this.props.onSubmit(values);
+		}
 	}
 
 	constructor(props) {
 		super(props);
-		this.state = {
-			valid: false,
-			fields: new Set(),
-		};
-
-		this.submit = this.submit.bind(this);
-		this.validate = this.validate.bind(this);
-		this.addToValidate = this.addToValidate.bind(this);
+
+		const { settings: { nameFieldRegistrationForm, emailFieldRegistrationForm, allowSwitchingDepartments },
+			departments } = props;
+
+		if (nameFieldRegistrationForm) {
+			this.state.name = { value: '' };
+		}
+
+		if (emailFieldRegistrationForm) {
+			this.state.email = { value: '' };
+		}
+
+		if (allowSwitchingDepartments && departments && departments.length > 0) {
+			this.state.department = { value: '' };
+		}
+
+		this.validateAll();
 	}
 
-	componentDidMount() {
-		this.validate();
+	componentWillReceiveProps({ settings: { nameFieldRegistrationForm, emailFieldRegistrationForm, allowSwitchingDepartments },
+		departments }) {
+		const hasName = nameFieldRegistrationForm;
+		const hasEmail = emailFieldRegistrationForm;
+		const hasDepartment = allowSwitchingDepartments && departments && departments.length > 0;
+
+
+		if (hasName && !this.state.name) {
+			this.setState({ name: { value: '' } });
+		} else if (!nameFieldRegistrationForm) {
```

Comment:
It's no big deal, but looking at the pattern below, you might want to test the `hasName` variable instead of `nameFieldRegistrationForm`.